errors/error.go

```diff
@@ -0,0 +1,373 @@
+		prev.Kind = ""
+	}
+
+	emptyRange := hcl.Range{}
+	if e.FileRange == emptyRange {
+		e.FileRange = prev.FileRange
+	}
+	if prev.FileRange == e.FileRange {
+		prev.FileRange = hcl.Range{}
+	}
+	if prev.Stack == e.Stack {
+		prev.Stack = nil
+	}
+	if prev.Description == e.Description {
+		prev.Description = ""
+	}
+	if prev.isEmpty() {
+		e.Err = prev.Err
+	}
+
+	return e
+}
+
+// isEmpty tells if all fields of this error are empty.
+// Note that e.Err is the underlying error hence not checked.
+func (e *Error) isEmpty() bool {
+	return e.FileRange == hcl.Range{} && e.Kind == "" && e.Description == "" && e.Stack == nil
+}
+
+func (e *Error) error(verbose bool) string {
+	if e.isEmpty() {
+		if e.Err != nil {
+			return e.Err.Error()
+		}
+		return ""
+	}
+
+	var errParts []string
+	for _, arg := range []interface{}{
+		e.FileRange,
+		e.Kind,
+		e.Description,
+		e.Stack,
+		e.Err,
+	} {
+		emptyRange := hcl.Range{}
+		switch v := arg.(type) {
+		case hcl.Range:
+			if v != emptyRange {
+				if verbose {
+					errParts = append(errParts,
+						fmt.Sprintf("filename=%q, start line=%d, start col=%d, "+
+							"start byte=%d, end line=%d, end col=%d, end byte=%d",
+							v.Filename,
+							v.Start.Line, v.Start.Column, v.Start.Byte,
+							v.End.Line, v.End.Column, v.End.Byte),
+					)
+				} else {
+					errParts = append(errParts, v.String())
+				}
+			}
+		case Kind:
+			if v != "" {
+				errParts = append(errParts, string(v))
+			}
+		case string:
+			if v != "" {
+				errParts = append(errParts, v)
+			}
+		case StackMeta:
+			if v != nil {
+				if verbose {
+					errParts = append(errParts,
+						fmt.Sprintf("at stack (name=%q, path=%q, desc=%q)",
+							v.Name(), v.Path(), v.Desc(),
+						))
+				} else {
+					errParts = append(errParts, fmt.Sprintf("at stack %q", v.Path()))
+				}
+			}
+		case error:
+			if v != nil {
+				errParts = append(errParts, v.Error())
+			}
+		case nil:
+			// ignore nil values
+		default:
+			panic(fmt.Errorf("unexpected errors.E type: %+v", arg))
+		}
+	}
+
+	return strings.Join(errParts, separator)
+}
+
+// Error returns the error message.
+func (e *Error) Error() string {
+	return e.error(false)
+}
+
+// Detailed returns a detailed error message.
+func (e *Error) Detailed() string {
+	return e.error(true)
+}
+
+// Is implements errors.Is interface.
+func (e *Error) Is(target error) bool {
+	return Is(e, target)
+}
+
+// IsKind tells if err is of kind k.
+// It returns false if err is nil or not an *errors.Error.
+// It also recursively checks if any underlying error is of kind k.
+func IsKind(err error, k Kind) bool {
+	if err == nil {
+		return false
+	}
+	e, ok := err.(*Error)
+	if !ok {
+		return false
+	}
+	if e.Kind != "" && e.Kind == k {
+		return true
+	}
+	return IsKind(e.Err, k)
+}
+
+// Is tells if err (or any of its underlying errors) matches target.
+// It works with any error but if comparing *Error type it uses the Kind field,
+// otherwise it fallback to standard errors.Is().
```

Comment:
```suggestion
// It works with any error but when comparing with an errors.Error type it will try to match the following fields:
// - Kind
// - Description
// - Stack
// - FileRange
// - Err
// Any fields absent (empty) on the target error are ignored even if they exist on err (partial match).
// If the target error is no an error.Error it fallbacks to standard errors.Is().
```